### PR TITLE
Parse the standard library

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Fix bugs in `TreeToIr` (rewrite) and parser. Implement more undocumented features in parser. Emulate some old parser bugs and quirks for compatibility.

Changes in libs:
- Fix some bugs.
- Clean up some odd syntaxes that the old parser translates idiosyncratically.
- Constructors are now required to precede methods.

### Important Notes

Out of 221 files:
- 215 match the old parser
- 6 contain complex types the old parser is known not to handle correctly

So, compared to the old parser, the new parser parses 103% of files correctly.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala [hunk 1]
```diff
@@ -6373,9 +6376,9 @@ object IR {
 
       /** @inheritdoc */
       override def toString: String =
+        //|doc = $doc,
```
Comment (kazcw) [inline engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala]: For testing; will remove change before merge.

File: engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala [hunk 2]
```diff
@@ -6474,9 +6477,9 @@ object IR {
 
       /** @inheritdoc */
       override def toString: String =
+        //|doc = ${doc.replaceAll("\n", "\\$")},
```
Comment (kazcw) [inline engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala]: For testing; will remove change before merge.

**Review by jdunkerley — APPROVED** [review]
Enso files checked. Thanks for these fixes.
One question do we need to have constructors first now before methods?

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 3]
```diff
@@ -263,6 +170,25 @@ case List(expr) => Some(expr)
         );
         yield cons(binding, appendTo);
       }
+      case Tree.ForeignFunction fn when fn.getBody() instanceof Tree.TextLiteral body -> {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:173]: Foreign function support. Great.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 4]
```diff
@@ -276,71 +202,16 @@ case List(expr) => Some(expr)
...
           reference,
           nil(),
-          translateExpression(a.getExpr(), false),
+          translateExpression(a.getExpr()),
           getIdentifiedLocation(a),
           meta(), diag()
         );
         yield cons(binding, appendTo);
       }
-    /*
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:285]: I was keeping these _scala code comments_ in order to know what the old parse did. Removing them is a signal we no longer need to study the old parser. That is great.

**Comment by kazcw** [comment]
> Enso files checked. Thanks for these fixes. One question do we need to have constructors first now before methods?

@jdunkerley When I was discussing the design for type definitions with him, @wdanilo told me that would be a requirement; however I noticed that when constructors don't precede methods, it looks like an intentional choice to place a private constructor after public methods. I have asked Wojciech if I should modify the design to allow it.

**Comment by kazcw** [comment]
(I meant to open this as a draft--I want to fix the 1 remaining IR difference before merge.)

File: engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala (hunk 2 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala]: Understandable. I have a similar change on #3611 branch as well.

File: engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java [hunk 5]
```diff
@@ -138,15 +138,6 @@ public void testIsDigitWithSpaces() throws Exception {
     """);
   }
 
-  @Test
-  @Ignore
-  public void testIsDigitWithoutSpaces() throws Exception {
```
Comment (JaroslavTulach) [inline engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java:143]: Right, this test is misleading as Wojciech wrote:

> `is_digit=character -> 42` is just `(is_digit=character) -> 42`, which is a lambda with named argument

the old parser doesn't understand that.

File: engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java [hunk 6]
```diff
@@ -405,7 +396,6 @@ public void testExportFrom() throws Exception {
   }
 
   @Test
-  @Ignore // wrong order of exported symbols
```
Comment (JaroslavTulach) [inline engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java:408]: Thank you.

File: lib/rust/parser/debug/LoadParser/LoadParser.java [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+    @Override
+    public void close() throws Exception {
+        parser.close();
+    }
+
+    public static void main(String[] args) throws Exception {
+        var root = new File(".").getAbsoluteFile();
+        try (LoadParser checker = new LoadParser(root)) {
+            checker.scan("distribution");
+            checker.scan("test");
```
Comment (JaroslavTulach) [inline lib/rust/parser/debug/LoadParser/LoadParser.java:55]: Yup, `test` is the right directory: https://github.com/enso-org/enso/pull/3611/commits/466ab24a8cb77172708b163add02670e7399ef13

File: lib/rust/parser/debug/tests/parse.rs [hunk 8]
```diff
@@ -207,14 +213,18 @@ fn type_operator_methods() {
         "type Foo",
         "    + : Foo -> Foo -> Foo",
         "    + self b = b",
+        "    Foo.+ : Foo",
+        "    Foo.+ self b = b",
```
Comment (JaroslavTulach) [inline lib/rust/parser/debug/tests/parse.rs:217]: Great.

**Review by JaroslavTulach — APPROVED** [review]
Thanks a lot for your magnificent work and moving forward both the Rust as well as Java side of the parser in such a tremendous way!

Btw. as a bonus you will have to update `ParseStdLibTest` as it [currently expects 67 additional failures](https://github.com/enso-org/enso/actions/runs/3327310241/jobs/5504911692#step:10:4250)...

**Comment by jdunkerley** [comment]
> > Enso files checked. Thanks for these fixes. One question do we need to have constructors first now before methods?
> 
> @jdunkerley When I was discussing the design for type definitions with him, @wdanilo told me that would be a requirement; however I noticed that when constructors don't precede methods, it looks like an intentional choice to place a private constructor after public methods. I have asked Wojciech if I should modify the design to allow it.

Happy to not do it if that is a requirement - just not one I knew was there!

**Comment by wdanilo** [comment]
Why this PR moves the `# PRIVATE` section above other sections?

File: engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala [hunk 9]
```diff
@@ -736,6 +736,9 @@ object IR {
             s"""
             |IR.Module.Scope.Import.Module(
             |name = $name,
+            |rename = $rename,
+            |onlyNames = $onlyNames,
+            |hiddenNames = $hiddenNames,
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala:741]: `isAll` is missing in the dump. Consider picking https://github.com/enso-org/enso/pull/3611/commits/34f49f299ca1ac2b3604c486338a3f6a1ffe9cca up.

File: engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java [hunk 10]
```diff
@@ -429,6 +419,27 @@ public void testTextLiteralWithEscape() throws Exception {
...
+  @Test
+  @Ignore
+  public void testRawBlockLiteral() throws Exception {
+    // mimics TextTest
+    parseTest("""
+    x = \"\"\"
+        Foo
+        Bar
+          Baz
+    """);
```
Comment (JaroslavTulach) [inline engine/runtime/src/test/java/org/enso/compiler/EnsoCompilerTest.java:431]: I am able to fix this one with:
```diff
@@ -858,7 +883,8 @@ final class TreeToIr {
     return new IR$Literal$Text(value, getIdentifiedLocation(txt), meta(), diag());
   }
   String buildTextConstant(Iterable elements, boolean stripComments) {
-    StringBuilder sb = new StringBuilder();
+    var sb = new StringBuilder();
+    var lines = 0;
     for (var t : elements) {
       switch (t) {
         case TextElement.Section s -> {
@@ -867,13 +893,20 @@ final class TreeToIr {
             // Reproduce an AstToIr bug for testing.
             text = text.replaceAll("#[^\n\r]*", "");
           }
+          if (text.endsWith("\n")) {
+              lines++;
+          }
           sb.append(text);
         }
         case TextElement.Escape e -> sb.appendCodePoint(e.getToken().getValue());
         default -> throw new UnhandledEntity(t, "buildTextConstant");
       }
     }
-    return sb.toString();
+    var text = sb.toString();
+    if (text.endsWith("\n") && lines > 1) {
+        text = text.substring(0, text.length() - 1);
+    }
+    return text;
   }
```
but I don't like my fix to commit it without your review, @kazcw.

**Comment by JaroslavTulach** [comment]
The CI is [failing](https://github.com/enso-org/enso/actions/runs/3344789612/jobs/5551568879#step:10:6119) with:
```
In module enso_dev.Tests.Semantic.Error_Spec:
Compiler encountered warnings:
Error_Spec.enso[238:74-238:80]: Unused function argument handler.
Compiler encountered errors:
Error_Spec.enso[238:82-238:84]: The name `err` could not be found.
Error_Spec.enso[239:18-239:20]: The name `err` could not be found.
Error_Spec.enso[240:18-240:20]: The name `err` could not be found.
Error_Spec.enso[240:55-240:57]: The name `err` could not be found.
Aborting due to 4 errors and 1 warnings.
Execution finished with an error: Compilation aborted due to errors.
```
Resolved by [e1ed61a](https://github.com/enso-org/enso/pull/3830/commits/e1ed61a427d0792be30c6e7ce870909230861395), as [Engine (linux) job](https://github.com/enso-org/enso/actions/runs/3356900991/jobs/5562319331) shows.

File: engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala (hunk 2 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala]: Removed in 5da58dc09.

File: engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/scala/org/enso/compiler/core/IR.scala]: Removed in 5da58dc09

**Comment by JaroslavTulach** [comment]
If the PR remains _green_, then there is no ready why not integrate it. It is a major leap forward.

**Pull request merged**